submit_cgap/scripts/submit_ontology.py

```diff
@@ -0,0 +1,41 @@
+import argparse
+
+from dcicutils.common import APP_FOURFRONT
+from ..submission import submit_any_ingestion
+from ..utils import script_catch_errors
+
+
+EPILOG = __doc__
+
+
+def main(simulated_args_for_testing=None):
+    parser = argparse.ArgumentParser(  # noqa - PyCharm wrongly thinks the formatter_class is invalid
+        description="Submits a gene list",
+        epilog=EPILOG,
+        formatter_class=argparse.RawDescriptionHelpFormatter
+    )
+    parser.add_argument('ontology_filename', help='a file of ontology data')
+    parser.add_argument('--lab', '-l', '-L', help='lab identifier', default=None)
+    parser.add_argument('--award', '-a', help='award identifier', default=None)
+    parser.add_argument('--server', '-s', help="an http or https address of the server to use", default=None)
+    parser.add_argument('--env', '-e', help="a CGAP beanstalk environment name for the server to use", default=None)
+    parser.add_argument('--validate-only', '-v', action="store_true",
+                        help="whether to stop after validating without submitting", default=False)
+    args = parser.parse_args(args=simulated_args_for_testing)
+
+    with script_catch_errors():
+
+        return submit_any_ingestion(
+                ingestion_filename=args.ontology_filename,
+                ingestion_type='ontology',
+                lab=args.lab,
+                award=args.award,
```

Comment:
I assumed it was a permissioning issue, that it might be needed. I get that an admin might not need it, but is this command only intended to be runnable by admins? SubmitCGAP can't know who is/isn't. If it's possible that someone's role might give them permission to also run this, then it should at least have it as an option.

Reply:
@aschroed You probably understand the  permissioning here, so maybe you can help clarify that. (I'm as interested in intended permissioning as implemented permissioning, since those might diverge as David is implementing this in parallel and not quite done.)

Reply:
Ontology updates are only by admin or the system.  The enumerated available status values for this Item makes them uneditable by non-admin in any case.  Currently, the submitter properties show that the 'load_xl' user of the system submits and updates terms.   OntologyTerm schema lacks lab and award properties and therefore there is not 'attribution' tab for this item type. 

The original intent was to have this all completely automated as part of the system.  It has subsequently shown that automatic updates are not really possible and that manual review of the terms generated by the script for updates is generally needed for a variety of reasons usually out of our control.

That's why I am not sure if @willronchetti 's suggestion about incorporating generate_ontology into the ingestion per se will work.

Intended permissioning is to have tight control of who can submit or change ontology terms - particularly those added via processing of files/terms provided by the official ontology maintainers.  We have our own 4DN controlled vocabulary ontology that we are a bit more flexible with where we can add terms we need that don't exist in existing ontologies yet.  That said I am not against more flexibility and being able to identify the specific person that did an update could be useful though attribution for these items doesn't add much IMO.

Reply:
@aschroed Thanks for that info. So it sounds like we could add a role that a person had to have (as an alternative to admin) in order to be able to perform this action. But to be clear, do you think there's value requiring either of Lab and Award. They're easily enough made optional, but Award, at least, is hard to reconstruct after-the-fact, so I don't want to miss the chance of recoring things we might later need, and I can't evaluate that very well. Is this action ever award-specific? It sounds like not. Is the role ever award-specific? I _think_ it sounds like not, but independent of the philosophizing, those are my concrete needs at this point: just to know if we should require either of those, too. Later we could extend permissioning, and for now we could make the submission only get accepted for processing under if you're an admin. Does that sound right?

Reply:
@aschroed Also, though there ia secondary question of whether we shold process the review in two steps. We could start with a single step and then break it up, or we could hold this functionality until we have a two-step process. We could offer some way to review the intermediate results before submitting them. That would be unusual but not impossible, and then there could be a secondary step that commits the result. It might require a UI step for the review, or perhaps we could make a spreadsheet of information for download that the person could then do an approval action on. It's possible these items are better discussed on the ticket than on the PR, but it's OK, I'll move relevant parts of this discussion later. Since it started here, we might as well finish it out here.

Reply:
> So it sounds like we could add a role that a person had to have (as an alternative to admin) in order to be able to perform this action.

That would be nice - similar to how a 'submits_for' works in Submit4DN though of course since that is tied to Lab in the authorization step really not that similar given my next answer.

> for now we could make the submission only get accepted for processing under if you're an admin. Does that sound right?

Yes - at least for now.

> We could start with a single step and then break it up, or we could hold this functionality until we have a two-step process. We could offer some way to review the intermediate results before submitting them. That would be unusual but not impossible, and then there could be a secondary step that commits the result. 

It would be quite dangerous to do this in a single step without review.  Currently there are 2 separate scripts - one to generate the json to load and the other to load it.  I recently found a 'bug' in the generation script that was due to script updates that did not anticipate a particular non-common scenario.  And the ontology maintainers themselves make changes or add bugs that are independent of our code.  There is some validation in the generator script but what action is required when things pop up is not always straightforward.

I am not familiar with exactly how the ingester works but for practicality for 4DN ontology terms and I don't see this need be much different for SMAHT ontologies I think the term generation should remain separate from the loading - the generated file can be presumably uploaded to AWS manually if needed for subsequent ingestion.  Only commenting on dealing with ontology term management here.